Disallow function variants of in-place operators other than existing ones

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22720 Disallow function variants of in-place operators other than existing ones**

Related to #22707

xuhdev committed:
Change in-place operators to method only (except those in torch.nn)

xuhdev commented:
Not sure whether the *elu functions in torch.nn require the same treatment (and they seem requiring special care for changing to method only). Should probably also add checks during the generation to ensure future compliance.

xuhdev committed:
Update on "[WIP] Change in-place operators to method only (except those in torch.nn)"

Change in-place operators to method only (except those in torch.nn)

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev commented:
CI blocked by https://github.com/pytorch/vision/pull/1107

xuhdev committed:
Update on "[WIP] Change in-place operators to method only (except those in torch.nn)"

Change in-place operators to method only (except those in torch.nn)

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

gchanan commented:
does anyone (@suo?) know if this change could break serialization (i.e. do we serialize models in such a way that they would call the function variants?).

gchanan commented:
Patch looks good.  I'd want to do a pass and figure out what this means for BC.

One quick thought: Just from a quick look, I don't see why we need copy_sparse_to_sparse_ exposed at all, it looks like implementations can just call it.

xuhdev commented:
CI waiting for #22754

xuhdev commented:
I also don't think copy_sparse_to_sparse_ should be exposed. But let's deal with that in another PR, because it looks like internal code depends on the fact that this function is generated from native_function.yaml

gchanan commented:
I looked through some of these functions and they've existed for awhile, so there are BC concerns.  My take here is if we don't JIT serialize these, then I would add deprecation warnings to the methods (would involve codegen changes) in order to not just break our users for a minor benefit.

xuhdev commented:
> I looked through some of these functions and they've existed for awhile, so there are BC concerns. My take here is if we don't JIT serialize these, then I would add deprecation warnings to the methods (would involve codegen changes) in order to not just break our users for a minor benefit.

Would you mind tagging #22707 or this as "need-discussion"? We can make a decision about how to deal with these functions before diving into the code...

gchanan commented:
> Would you mind tagging #22707 or this as "need-discussion"? We can make a decision about how to deal with these functions before diving into the code...

done.

xuhdev committed:
Update on "[WIP] Change in-place operators to method only (except those in torch.nn)"

Deprecate function versions of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "[WIP] Deprecate function versions of in-place operators."

Deprecate function versions of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "[WIP] Deprecate function versions of in-place operators."

Deprecate function versions of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "[WIP] Deprecate function versions of in-place operators."

Deprecate function versions of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "[WIP] Deprecate function versions of in-place operators."

Deprecate function versions of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev commented:
@pytorchbot retest this please

xuhdev committed:
Update on "Deprecate function versions of in-place operators. (not those in the nn module)"

Deprecate function versions of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "Deprecate function versions of in-place operators. (not those in the nn module)"

Deprecate function versions of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "Deprecate function versions of in-place operators. (not those in the nn module)"

Deprecate function versions of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev commented:
@gchanan This PR should be ready now. Feel free to review.

xuhdev committed:
Update on "Deprecate function versions of in-place operators. (not those in the nn module)"

Deprecate function versions of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -702,6 +706,17 @@ def process_function(name, declarations):
+        if not is_module and not has_self and name.endswith('_') and not name.startswith('_'):
+            # After the deprecated functions are removed, replace this block with an AssertionError to ensure that
+            # future in-place operators won't be exposed as functions.
+            env['deprecation_message'] = ('PyErr_WarnEx(PyExc_DeprecationWarning, '
```
use TORCH_WARN instead.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -83,18 +83,22 @@
 - func: dropout(Tensor input, float p, bool train) -> Tensor
 
 - func: dropout_(Tensor(a!) self, float p, bool train) -> Tensor(a!)
+  variant: function, method
```
I'd prefer to have an indication in native_functions that something is wrong; it seems totally feasible that someone would copy-paste one of these, not happen to call the function variant and not get a warning.

I'm not sure if this would work because not all of these are deprecated, but I'd consider:
1) adding an annotation, e.g. "deprecated_method: True"
2) If you have one of these bad methods that don't have a deprecated_method tag (and isn't in NN?) you error out.

Is that feasible?

xuhdev replied:
Do you think it would be simpler if we report an error in `gen_python_functions.py` (where TORCH_WARN is suggested), if the function is not one of the currently existing ones? I don't think anyone should add a new deprecated inplace operator that can be used as a function at this point (we are kind of inconsistent about whether an inplace operator has a function variant at this point).

And the time window to add a new such operator is probably short because we can actually come back and revisit this issue in two weeks, assuming a new release comes out before the end of July :)

gchanan replied:
a whitelist also seems fine, as long as we still TORCH_WARN for those.

gchanan reviewed (commented):
looks good!  One suggestion and one question.

xuhdev committed:
Update on "Deprecate function versions of in-place operators. (not those in the nn module)"

Deprecate function variants of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "Deprecate function versions of in-place operators. (not those in the nn module)"

Deprecate function variants of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev commented:
Do we also wanna do this for functions in the nn module in the future?

gchanan commented:
I don't know, we don't really have a solution for that (I guess some functions use `inplace=True` but that seems weird...).

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -702,6 +706,26 @@ def process_function(name, declarations):
 
+        # In-place operator, should be method only. Deprecate it.
+        if not is_module and not has_self and name.endswith('_') and not name.startswith('_'):
+            deprecated_functions = ['abs_', 'acos_', 'addmv_', 'alpha_dropout_', 'as_strided_', 'asin_', 'atan_',
```
ya, I don't think putting a big list here is a good idea.

For example, I don't see how we can deprecate a bunch of these -- our nn functional code calls a bunch of these, even if they aren't in the "correct" nn namespace (check all the dropouts, for example).

xuhdev replied:
I believe that the use of all of these function variants have been eliminated, at least in PyTorch codebase (No compilation error if we simply remove them from the scene).

Where would you like to put the list? This list is not supposed to grow anyway and we only hide them here temporarily.

What is your suggesetion for handling misplaced nn functions?

gchanan replied:
> I believe that the use of all of these function variants have been eliminated, at least in PyTorch codebase (No compilation error if we simply remove them from the scene).

What about something like this?
https://github.com/pytorch/pytorch/blob/eb76b7a564121c5fede749ad7d0a36f2b61a0a95/torch/nn/functional.py#L800

xuhdev replied:
OK, fair enough. I reverted all changes of all nn functions that are misplaced in `torch.` (in this context, `*dropout` functions). We can probably decide what we should do with those functions in the future. Could you review again?

xuhdev replied:
Do you in general treat `relu_` as an nn function? If so, I'll also remove its deprecation as well.

xuhdev replied:
@gchanan This PR should be ready now. Feel free to review. Thanks!

xuhdev committed:
Update on "Deprecate function variants of in-place operators. (not those in the nn module)"

Deprecate function variants of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "Deprecate function variants of in-place operators. (not those in the nn module)"

Deprecate function variants of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "Deprecate function variants of in-place operators. (not those in the nn module)"

Deprecate function variants of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "Deprecate function variants of in-place operators. (not those in the nn module)"

Deprecate function variants of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "Deprecate function variants of in-place operators. (not those in the nn module)"

Deprecate function variants of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "Deprecate function variants of in-place operators. (not those in the nn module)"

Deprecate function variants of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

VitalyFedyunin reviewed (changes requested):
Please exclude `third_party/onnx` update

VitalyFedyunin commented on tools/autograd/gen_python_functions.py:
```diff
@@ -702,6 +706,26 @@ def process_function(name, declarations):
+
+            # After the deprecated functions are removed, replace this block with an AssertionError to ensure that
+            # future in-place operators won't be exposed as functions.
+            env['deprecation_message'] = ('TORCH_WARN("In-place functions such as torch.{name} are deprecated '
```
Please define template as global variable

VitalyFedyunin commented on tools/autograd/gen_python_functions.py:
```diff
@@ -702,6 +706,26 @@ def process_function(name, declarations):
 
+        # In-place operator, should be method only. Deprecate it.
+        if not is_module and not has_self and name.endswith('_') and not name.startswith('_'):
+            deprecated_functions = ('dropout_', 'feature_dropout_', 'alpha_dropout_', 'feature_alpha_dropout_',
```
Please move to global var and add comment explaining it's meaning

VitalyFedyunin commented on tools/autograd/gen_python_functions.py:
```diff
@@ -702,6 +706,26 @@ def process_function(name, declarations):
+                                          'and will be removed in the next release. '
+                                          'Please use their in-place method counterparts (Tensor.{name}) instead.");'
+                                          ).format(name=name)
+            env['deprecated'] = True
```
Unused

xuhdev committed:
Update on "Deprecate function variants of in-place operators. (not those in the nn module)"

Deprecate function variants of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

xuhdev committed:
Update on "Deprecate function variants of in-place operators. (not those in the nn module)"

Deprecate function variants of in-place operators.

gh-metadata: pytorch pytorch 22720 gh/xuhdev/9/head

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -66,6 +68,7 @@
 PY_VARIABLE_METHOD_NOARGS = CodeTemplate("""\
 static PyObject * ${pycname}(PyObject* self_, PyObject* args)
 {
+  ${deprecation_message}
```
hmm this is interesting, because removing these won't actually let us get rid of the methods, because this is only affecting the python API (and it currently must at least include this because the python functions can call C++ methods, so you can't put the deprecation warning in the C++ methods).

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -165,6 +168,20 @@
+DEPRECATED_INPLACE_FUNCTIONS = ('abs_', 'acos_', 'addmv_', 'as_strided_', 'asin_', 'atan_', 'ceil_', 'celu_',
+                                'clamp_', 'clamp_max_', 'clamp_min_', 'cos_', 'cosh_', 'detach_',
+                                'embedding_renorm_', 'erf_', 'erfc_', 'exp_', 'expm1_', 'fill_', 'floor_', 'frac_',
+                                'index_put_', 'log10_', 'log1p_', 'log2_', 'log_', 'neg_', 'reciprocal_', 'relu_',
```
how did you determine this list?  For example, it looks like rrelu_ is used in nn/functional, but it's not in the EXEMPTED list.

Did you check that none of these methods are actually used?

xuhdev replied:
I'm still unsure yet what to do with `relu`, `rrelu`, and `celu`. These look like nn functions but they are under torch, and their positioning is kind of vague. For all others, I add them to deprecation list because they are in-place and semantically shouldn't be in nn.

gchanan reviewed (changes requested):
I'm pretty dubious about this change, because I think there are some pre-reqs we'd need to do first.

1) change python function bindings to call functions, not methods, and ensure there is no performance regression.
2) make sure we aren't introducing any warnings in our own code
3) warn at the C++ level, not just python.

xuhdev commented:
> I'm pretty dubious about this change, because I think there are some pre-reqs we'd need to do first.
> 
>     1. change python function bindings to call functions, not methods, and ensure there is no performance regression.
> 
>     2. make sure we aren't introducing any warnings in our own code
> 
>     3. warn at the C++ level, not just python.

How about this: before we come back to think about how to deal with the deprecation, I'm gonna create another PR (or convert this PR) to stop additional inplace operators from being declared as function variants first (so as to prevent the situation from getting worse). Do you think it is sufficient to do this in tools/autograd/gen_python_functions.py by emitting an error if a new in-place operator is declared as function variant? It should look very similar to the current shape except that there is no deprecation message.

And Point 1 benefits not only this single deprecation purpose, but also the consistency within function and method variants.

VitalyFedyunin commented:
> Do you think it is sufficient to do this in tools/autograd/gen_python_functions.py by emitting an error if a new in-place operator is declared as function variant? It should look very similar to the current shape except that there is no deprecation message.

Sounds good to me, as soon as we have detailed error message at codegen step.

xuhdev committed:
Update on "Deprecate function variants of in-place operators. (not those in the nn module)"


Related to #22707

xuhdev committed:
Update on "Deprecate function variants of in-place operators. (not those in the nn module)"


Related to #22707

xuhdev commented:
@VitalyFedyunin  Updated

VitalyFedyunin commented:
Hi! Can you please rebase.

xuhdev committed:
Update on "Disallow function variants of in-place operators other than existing ones"


Related to #22707

[ghstack-poisoned]

xuhdev commented:
@VitalyFedyunin Sorry for the delay. I've rebased now

xuhdev committed:
Update on "Disallow function variants of in-place operators other than existing ones"


Related to #22707

[ghstack-poisoned]

xuhdev commented:
@VitalyFedyunin Mind looking into this :)
